Tools/cases_generator/opcode_metadata_generator.py

```diff
@@ -332,22 +335,30 @@ def generate_expansion_table(analysis: Analysis, out: CWriter) -> None:
             assert (
                 len(instr2.parts) == 1
             ), f"{name2} is not a good superinstruction part"
-            expansions.append((instr1.parts[0].name, OPARG_KINDS["OPARG_TOP"], 0))
-            expansions.append((instr2.parts[0].name, OPARG_KINDS["OPARG_BOTTOM"], 0))
+            expansions.append((instr1.parts[0].name, "OPARG_TOP", 0))
+            expansions.append((instr2.parts[0].name, "OPARG_BOTTOM", 0))
         elif not is_viable_expansion(inst):
             continue
         else:
             for part in inst.parts:
                 size = part.size
-                if part.name == "_SAVE_RETURN_OFFSET":
-                    size = OPARG_KINDS["OPARG_SAVE_RETURN_OFFSET"]
                 if isinstance(part, Uop):
+                    fmt = "0"
+                    if part.name == "_SAVE_RETURN_OFFSET":
+                        fmt = "OPARG_SAVE_RETURN_OFFSET"
+                    elif part.caches:
+                        fmt = str(part.caches[0].size)
                     # Skip specializations
                     if "specializing" in part.annotations:
                         continue
                     if "replaced" in part.annotations:
-                        size = OPARG_KINDS["OPARG_REPLACED"]
-                    expansions.append((part.name, size, offset if size else 0))
+                        fmt = "OPARG_REPLACED"
+                    expansions.append((part.name, fmt, offset))
+                    if len(part.caches) > 1:
+                        internal_offset = 0
+                        for cache in part.caches[:-1]:
+                            internal_offset += cache.size
+                        expansions.append((part.name, f"OPERAND1_{part.caches[-1].size}", offset+internal_offset))
```

Comment:
Likewise, this might benefit from an extra comment:

```suggestion
                    # Add an expansion for the extra cache space, if there's more than one.
                    if len(part.caches) > 1:
                        internal_offset = 0
                        for cache in part.caches[:-1]:
                            internal_offset += cache.size
                        expansions.append((part.name, f"OPERAND1_{part.caches[-1].size}", offset+internal_offset))
```

Reply:
Also, to be sure I've understood correctly, this code is adding an expansion for the part's last cache entry.  Is that right?

Relatedly, the special-casing for "OPERAND1_" implies to me that operand 0 is the regular case and that there are only ever at most 2 operands.  Is that right?

With that in mind, this code seems like it's halfway between accommodating a dedicated second operand/cache entry and accommodating an arbitrary number of extra ones.  It might be more clear to make the dedicated second one more explicit:

```suggestion
                    # Add an expansion for a second cache entry, if there is one.
                    if len(part.caches) > 1:
                        assert len(part.caches) == 2, len(part.caches)
                        firstsize = part.caches[0].size
                        secondsize = part.caches[-1].size
                        expansions.append((part.name, f"OPERAND1_{secondsize}", offset+ firstsize))
```

...or perhaps I've misunderstood what's going on (a real possibility)?

(Also, what's the likelihood that in the future there might be uops with a third cache entry or more?)

Reply:
> there are only ever at most 2 operands. Is that right?

Yes. It is unlikely, but not impossible, that we will ever add a third.

Reply:
There can be unused cache entries between the used ones, so we need to iterate over them to get the offsets.
E.g.
```
op(_LOAD_GLOBAL_MODULE, (version/1, unused/1, index/1 -- res))
```

Reply:
Thanks for explaining.  I have a follow-up comment that I'll leave in my next review.